### PR TITLE
Check if the service is not found during reconcileDelete

### PR DESCRIPTION
**Pull request by fmount**
This patch adds an error check to the glance-operator during reconcileDelete: if it's an error and it's not found we can assume the service has been deleted, otherwise we must fail. The same logic has been added to the cinder operator, and now we have a well established pattern to explicitly check and clean the related resources.

Signed-off-by: Francesco Pantano <fpantano@redhat.com>

### CODE REVIEW

**Review by abays — APPROVED** [review]
/lgtm

**Comment by fmount** [comment]
@abays I'm wondering if (since we're fixing this behavior), we'd like to reflect the same change on the other operators.
Here what I found from a quick look:

- https://github.com/openstack-k8s-operators/neutron-operator/blob/master/controllers/neutronapi_controller.go#L195
- https://github.com/openstack-k8s-operators/placement-operator/blob/master/controllers/placementapi_controller.go#L205
- https://github.com/openstack-k8s-operators/ironic-operator/blob/master/controllers/ironicapi_controller.go#L249

**Comment by abays** [comment]
> @abays I'm wondering if (since we're fixing this behavior), we'd like to reflect the same change on the other operators. Here what I found from a quick look:
> 
>     * https://github.com/openstack-k8s-operators/neutron-operator/blob/master/controllers/neutronapi_controller.go#L195
> 
>     * https://github.com/openstack-k8s-operators/placement-operator/blob/master/controllers/placementapi_controller.go#L205
> 
>     * https://github.com/openstack-k8s-operators/ironic-operator/blob/master/controllers/ironicapi_controller.go#L249

I made a similar remark elsewhere.  It seems to me that we should indeed be consistent with this pattern across all the operators.

File: controllers/glanceapi_controller.go [hunk 1]
```diff
@@ -207,10 +207,16 @@ func (r *GlanceAPIReconciler) reconcileDelete(ctx context.Context, instance *gla
...
 	// It's possible to get here before the endpoints have been set in the status, so check for this
 	if instance.Status.APIEndpoints != nil {
-		ks, _ := keystone.GetKeystoneServiceWithName(ctx, helper, glance.ServiceName, instance.Namespace)
+		ks, err := keystone.GetKeystoneServiceWithName(ctx, helper, glance.ServiceName, instance.Namespace)
+		if err != nil {
+			if k8s_errors.IsNotFound(err) {
+				return ctrl.Result{}, nil
+			}
+			return ctrl.Result{}, err
+		}
```
Comment (stuggi) [inline controllers/glanceapi_controller.go:216]: I just returned from a longer PTO break so maybe I miss something as my brain us slowly coming into work mode :) 

I am not sure I understand why we need this here when we do the same check in ksSvc.Delete() [1]. and return nil if it does not exist. Isn't it correct in that case to proceed and with the remaining delete tasks.

[1] https://github.com/openstack-k8s-operators/keystone-operator/blob/master/pkg/external/funcs.go#L211

Comment (stuggi) [reply controllers/glanceapi_controller.go:216]: Now I got it, thanks @abays , to not have to use the service definition for ks.
I think there is still an issue. If we return `return ctrl.Result{}, nil` in the k8s_errors.IsNotFound case we won't proceed and remove the finalizer later.

Comment (fmount) [reply controllers/glanceapi_controller.go:216]: @stuggi just a quick question: is `k8s_errors.IsNotFound(err)` assumes that the service is not found? (hence we can return earlier?)

Comment (stuggi) [reply controllers/glanceapi_controller.go:216]: yes `k8s_errors.IsNotFound(err)` here assumes that the glanceapi keystoneservice CR is not there, so you want to skip the `ksSvc.Delete()` but still proceed the remaining parts of the func for cleanup. Until there are further cleanup steps the removal of the finalizer on the CR `controllerutil.RemoveFinalizer(instance, helper.GetFinalizer())` .

Comment (abays) [reply controllers/glanceapi_controller.go:216]: @fmount My bad for not spotting this in the similar PR for Cinder.  I have a follow-up PR open there to address the concern Martin is sharing

Comment (fmount) [reply controllers/glanceapi_controller.go]: @abays np, let's see if this version is more aligned with your fix on the cinder side, then I guess we can move on.

**Review by abays — APPROVED** [review]
/lgtm

**Pull request merged**